controllers/dataprotection/backuprepo_controller.go

```diff
@@ -0,0 +1,831 @@
+		content, err := renderTemplate("secret", provider.Spec.CSIDriverSecretTemplate, renderCtx)
+		if err != nil {
+			_ = updateCondition(reqCtx.Ctx, r.Client, repo, condType,
+				metav1.ConditionFalse, ReasonBadSecretTemplate, err.Error())
+			return fmt.Errorf("failed to render secret template: %w", err)
+		}
+		secretStringData := map[string]string{}
+		if err = yaml.Unmarshal([]byte(content), &secretStringData); err != nil {
+			_ = updateCondition(reqCtx.Ctx, r.Client, repo, condType,
+				metav1.ConditionFalse, ReasonBadSecretTemplate, err.Error())
+			return fmt.Errorf("failed to unmarshal secret content: %w", err)
+		}
+		secretData := make(map[string][]byte, len(secretStringData))
+		for k, v := range secretStringData {
+			secretData[k] = []byte(v)
+		}
+		secret.Data = secretData
+		return nil
+	}
+
+	secret := &corev1.Secret{}
+	secret.Name = repo.Status.GeneratedCSIDriverSecret.Name
+	secret.Namespace = repo.Status.GeneratedCSIDriverSecret.Namespace
+
+	// create the secret object if not exist.
+	// this function will retrieve the whole secret object
+	// when the object is existing.
+	created, err = createObjectIfNotExist(reqCtx.Ctx, r.Client, secret,
+		func() error {
+			secret.Labels = map[string]string{
+				dataProtectionBackupRepoKey: repo.Name,
+			}
+			secret.Annotations = map[string]string{
+				dataProtectionSecretTemplateMD5AnnotationKey: secretTemplateMD5,
+				dataProtectionTemplateValuesMD5AnnotationKey: templateValuesMD5,
+			}
+			if err := setSecretContent(secret); err != nil {
+				return err
+			}
+			if err := controllerutil.SetControllerReference(repo, secret, r.Scheme); err != nil {
+				_ = updateCondition(reqCtx.Ctx, r.Client, repo, condType,
+					metav1.ConditionUnknown, ReasonUnknownError, err.Error())
+				return fmt.Errorf("failed to set controller reference: %w", err)
+			}
+			return nil
+		})
+	if err != nil {
+		return false, fmt.Errorf("createObjectIfNotExist for secret %s failed: %w",
+			client.ObjectKeyFromObject(secret), err)
+	}
+	if created {
+		return true, nil
+	}
+
+	// check if the template or config changed, then update the secret
+	currSecretTemplateMD5 := secret.Annotations[dataProtectionSecretTemplateMD5AnnotationKey]
+	currTemplateValuesMD5 := secret.Annotations[dataProtectionTemplateValuesMD5AnnotationKey]
+	if currSecretTemplateMD5 != secretTemplateMD5 || currTemplateValuesMD5 != templateValuesMD5 {
+		patch := client.MergeFrom(secret.DeepCopy())
+		if err := setSecretContent(secret); err != nil {
+			return false, err
+		}
+		if secret.Annotations == nil {
+			secret.Annotations = make(map[string]string)
+		}
+		secret.Annotations[dataProtectionSecretTemplateMD5AnnotationKey] = secretTemplateMD5
+		secret.Annotations[dataProtectionTemplateValuesMD5AnnotationKey] = templateValuesMD5
+		err := r.Client.Patch(reqCtx.Ctx, secret, patch)
+		if err != nil {
+			return false, fmt.Errorf("failed to patch secret object %s: %w",
+				client.ObjectKeyFromObject(secret), err)
+		}
```

Comment:
There is no need to wrap the error, the information returned by apiserver is already very detailed